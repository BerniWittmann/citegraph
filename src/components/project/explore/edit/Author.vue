<template>
  <div class="project-explore-single-author-edit">
    <v-card class="pa-4">
      <v-card-title class="headline">
        <h2 class="display-5" v-once>{{ author.firstName }} {{ author.lastName }}</h2>
      </v-card-title>
      <v-container>
        <v-form ref="form" @submit.prevent="submit" class="mt-12" v-model="valid">
          <v-row justify="center">
            <v-col lg="6">
              <v-text-field
                :label="$t('project.explore.edit.author.first_name')"
                v-model="firstName"
                :rules="nameRules"
                required
                autofocus/>
            </v-col>
          </v-row>
          <v-row justify="center">
            <v-col lg="6">
              <v-text-field
                :label="$t('project.explore.edit.author.last_name')"
                v-model="lastName"
                :rules="nameRules"
                required
                autofocus/>
            </v-col>
          </v-row>
          <v-row class="my-10" justify="space-between">
            <v-col cols="auto">
              <v-btn rounded large color="warning" @click="back">
                <v-icon>mdi-menu-left</v-icon>
                {{ $t('project.explore.edit.back') }}
              </v-btn>
            </v-col>

            <v-col cols="auto">
              <v-btn
                rounded large
                :disabled="!valid"
                :loading="loading"
                color="success"
                type="submit"
              >{{ $t('project.explore.edit.submit') }}
              </v-btn>
            </v-col>
          </v-row>
        </v-form>
      </v-container>
    </v-card>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import Author from '@/models/paperEntities/author'

@Component
export default class AuthorEditPage extends Vue {
  valid: boolean = true
  loading: boolean = false
  firstName: string = ''
  lastName: string = ''

  nameRules: Array<Function> = [
    (v: string | undefined) => !!v || this.$root.$t('project.explore.edit.author.name_required'),
    (v: string | undefined) => (v && v.trim().length > 0) || this.$root.$t('project.explore.edit.author.name_not_empty')
  ]

  urlRules: Array<Function> = []

  get author (): Author {
    return this.$store.getters['paperEntities/activeEntity'] as Author
  }

  beforeMount (): void {
    this.firstName = this.author.firstName
    this.lastName = this.author.lastName
  }

  async submit (): Promise<void> {
    if (!this.valid) return
    this.loading = true
    await this.$store.dispatch('paperEntities/updateEntity', {
      params: {
        projectId: this.$route.params.projectId,
        entityType: this.$route.params.queryByType,
        id: this.$route.params.entityId
      },
      data: {
        firstName: this.firstName,
        lastName: this.lastName
      }
    })
    this.loading = false
    this.$router.back()
  }

  back (): void {
    this.$router.back()
  }
}
</script>
