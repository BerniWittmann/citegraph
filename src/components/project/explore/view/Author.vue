<template>
  <div class="project-explore-single-author">
    <layout>
      <template v-slot:title>
        <v-avatar color="primary" class="mr-5" size="62">
          <b class="white--text">{{ initials }}</b>
        </v-avatar>
        <h2 class="display-5">{{ author.firstName }} {{ author.lastName }}</h2>
      </template>
      <v-container>
        <v-row>
          <v-col :cols="3">
            <v-list>
              <v-subheader class="title">{{ $t('project.explore.view.author.metrics') }}</v-subheader>
              <v-list-item>
                <v-list-item-content>{{ $t('project.explore.view.author.count_records') }}:</v-list-item-content>
                <v-list-item-content class="align-end">{{ author.countRecords }}</v-list-item-content>
              </v-list-item>
            </v-list>
          </v-col>
        </v-row>
      </v-container>
    </layout>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import PaperEntityLayout from '@/layouts/PaperEntity.vue'
import Author from '@/models/paperEntities/author'

@Component({
  components: {
    Layout: PaperEntityLayout
  }
})
export default class AuthorViewPage extends Vue {
  get author (): Author {
    return this.$store.getters['paperEntities/activeEntity'] as Author
  }

  get initials (): string {
    return (this.author.firstName.substring(0, 1) + this.author.lastName.substring(0, 1)).toUpperCase()
  }
}
</script>
