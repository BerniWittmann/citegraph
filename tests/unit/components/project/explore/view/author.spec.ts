import { shallowMount } from '@vue/test-utils'
import { i18n } from '../../../../setupPlugins'

import AuthorViewPage from '@/components/project/explore/view/Author.vue'
import EmptySlotComponent from '../../../../EmptySlotComponent.vue'
import Author from '@/models/paperEntities/author'

describe('components/project/explore/view/Author.vue', () => {
  function getWrapper () {
    return shallowMount(AuthorViewPage, {
      i18n,
      mocks: {
        $store: {
          getters: {
            'paperEntities/activeEntity': new Author({ id: '1', firstName: 'Max', lastName: 'Mustermann', countRecords: 12 })
          }
        }
      },
      stubs: {
        layout: EmptySlotComponent
      }
    })
  }

  it('renders the page', () => {
    const wrapper = getWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders the name', () => {
    const wrapper = getWrapper()
    expect(wrapper.text()).toContain('Max Mustermann')
  })

  it('renders a data table for the records', () => {
    const wrapper = getWrapper()
    const table = wrapper.find('data-table-stub')
    expect(table.exists()).toBeTruthy()
    expect(table.props()).toEqual({
      belongsTo: '1',
      belongsToType: 'author',
      queryByType: 'record'
    })
  })
})
