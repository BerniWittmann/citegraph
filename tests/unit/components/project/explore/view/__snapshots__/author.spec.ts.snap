// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/project/explore/view/Author.vue renders the page 1`] = `
<div class="project-explore-single-author">
  <div class="empty-slot-component-stub">
    <v-avatar-stub color="primary" size="62" class="mr-5"><b class="white--text">MM</b></v-avatar-stub>
    <h2 class="display-5">Max Mustermann</h2>
    <v-container-stub tag="div">
      <v-row-stub tag="div">
        <v-col-stub lg="6" xl="3" tag="div">
          <v-list-stub tag="div" tile="true">
            <v-subheader-stub class="title">project.explore.view.author.metrics</v-subheader-stub>
            <v-list-item-stub activeclass="" tag="div">
              <v-list-item-content-stub>project.explore.view.author.count_records:</v-list-item-content-stub>
              <v-list-item-content-stub class="align-end">12</v-list-item-content-stub>
            </v-list-item-stub>
          </v-list-stub>
        </v-col-stub>
      </v-row-stub>
      <v-row-stub tag="div">
        <data-table-stub querybytype="record" belongsto="1" belongstotype="author"></data-table-stub>
      </v-row-stub>
    </v-container-stub>
  </div>
</div>
`;
